### PR TITLE
892 implement simple lct model

### LINKED ISSUE #892: Implement simple LCT model
### Feature description

- add lct model
- simple example
- related tests
- replace print by print_table

### Additional context

Copy code from #680 
Files:
- lct_secir without initialization, parameters_io
- example lct_secir
- test

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project.

### PR DESCRIPTION
**Pull request by lenaploetzke**
# Changes and Information

Please **briefly list the changes** (main added features, changed items, or corrected bugs) made:

- Added simple version of the lct secir model with related example and test. 

If need be, add additional information and what the reviewer should look out for in particular:

- Most important file is model.cpp
- simulation.h is comparable to memilio/compartments/simulation.h, see also #914.

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/SciCompMod/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/SciCompMod/memilio/wiki/Coding-guidelines)
- [ ] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added
- [ ] (For ABM development) Checked [benchmark results](https://github.com/SciCompMod/memilio/wiki/Agent-Based-Model-Development) and ran and posted a local test above from before and after development to ensure performance is monitored.

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes, code coverage and performance is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] On merge, add 2-5 lines with the changes (main added features, changed items, or corrected bugs) to the merge-commit-message. This can be taken from the **briefly-list-the-changes** above (best case) or the separate commit messages (worst case).

Closes #892.

### CODE REVIEW

File: cpp/models/lct_secir/README.md
Comment (mknaranja) [inline cpp/models/lct_secir/README.md:5]:
```suggestion
- Lena Plötzke, "Der Linear Chain Trick in der epidemiologischen Modellierung als Kompromiss zwischen gewöhnlichen und Integro-Differentialgleichungen", 2023. (https://elib.dlr.de/200381/, German only)
```

File: cpp/examples/lct_secir.cpp
Comment (mknaranja) [inline cpp/examples/lct_secir.cpp:32]:
```suggestion
    /** Simple example to demonstrate how to run a simulation using an LCT SECIR model. 
```

File: cpp/examples/lct_secir.cpp
Comment (mknaranja) [inline cpp/examples/lct_secir.cpp:33]:
```suggestion
    Parameters, initial values and subcompartments are not meant to represent a realistic scenario. */
```

**Review by HenrZu — CHANGES_REQUESTED** [review]
Nice, good work :)

**Comment by HenrZu** [comment]
@mknaranja I think this PR can lead to issues with PR #894. Do we wait until this is merged?

**Comment by mknaranja** [comment]
@HenrZu @lenaploetzke We should advance both pull requests with review (and adaptations) at the same time and for the one to be merged later we should combine forces to not put to much on the shoulder of the persons who comes second ;)

File: cpp/examples/lct_secir.cpp
Comment (lenaploetzke) [reply cpp/examples/lct_secir.cpp:24]: I am using ScalarType below, which is defined in config.h.

File: cpp/models/lct_secir/infection_state.h
Comment (lenaploetzke) [reply cpp/models/lct_secir/infection_state.h:111]: Yes, rewrote whole function.

File: cpp/models/lct_secir/model.h
Comment (lenaploetzke) [reply cpp/models/lct_secir/model.h:95]: Not working because m_N0 is calculated in the constructor using other input variables.

File: cpp/examples/lct_secir.cpp
Comment (HenrZu) [inline cpp/examples/lct_secir.cpp:90]:
```suggestion
    mio::TimeSeries<ScalarType> population_no_subcompartments = model.calculate_populations(result);
    population_no_subcompartments.print_table({"S", "E", "C", "I", "H", "U", "R", "D "}, 16, 8);
```

File: cpp/models/lct_secir/infection_state.h
Comment (HenrZu) [inline cpp/models/lct_secir/infection_state.h:110]: Intuitively, it would make more sense to me if the check_constraint() function is true when everything is ok. But if you want to leave it like this, you can simply resolve this comment.
´

File: cpp/models/lct_secir/infection_state.h
Comment (HenrZu) [inline cpp/models/lct_secir/infection_state.h:113]: I think a mio::log_warning here maybe useful. Or even a log_error as it does not make much sense to continue with invalid vectors

File: cpp/models/lct_secir/infection_state.h
Comment (HenrZu) [inline cpp/models/lct_secir/infection_state.h:145]: again. Maybe add a mio::log_warning msg

File: cpp/models/lct_secir/infection_state.h
Comment (HenrZu) [inline cpp/models/lct_secir/infection_state.h:169]:
```suggestion
     * @return Index of the first subcompartment for a vector with one entry per subcompartment or -1.
```

File: cpp/models/lct_secir/infection_state.h
Comment (HenrZu) [inline cpp/models/lct_secir/infection_state.h:194]: Something else I've just noticed. We have the public function check_constraints() here. In itself, I think the naming is good. But this could be a bit confusing with the existing check_constraints() functions for the parameters and population. I would recommend making the name more specific.

File: cpp/models/lct_secir/infection_state.h
Comment (HenrZu) [inline cpp/models/lct_secir/infection_state.h:106]: in general, if the function works as intended, we should return true :)

**Review by HenrZu — DISMISSED** [review]

File: cpp/models/lct_secir/infection_state.h
Comment (lenaploetzke) [reply cpp/models/lct_secir/infection_state.h:110]: Basically I agree with you here, but the function was implemented the same way in the IDE and OD models. That's why I wanted to keep the consistency here. Maybe this could be changed for all models, but given the description in the function documentation, I think it's actually okay.

File: cpp/models/lct_secir/infection_state.h
Comment (lenaploetzke) [reply cpp/models/lct_secir/infection_state.h:194]: My idea here was that the function should be similar to the ones from parameters and population. It actually has the same purpose, except that I call it in a different place to avoid calculations with invalid vectors.

File: cpp/examples/lct_secir.cpp
Comment (annawendler) [inline cpp/examples/lct_secir.cpp:40]: The notation infectionState is inconsistent with the remaining notation, maybe use infection_state?

File: cpp/models/lct_secir/infection_state.h
Comment (annawendler) [inline cpp/models/lct_secir/infection_state.h:54]: Where do you need this?

File: cpp/models/lct_secir/infection_state.h
Comment (annawendler) [inline cpp/models/lct_secir/infection_state.h:74]:
```suggestion
     * InfectionState class defines the possible InfectionState%s with the number of subcompartments for the LCT model.
```

File: cpp/models/lct_secir/infection_state.h
Comment (annawendler) [inline cpp/models/lct_secir/infection_state.h:88]:
```suggestion
     * InfectionState class defines the possible InfectionState%s with the number of subcompartments for the LCT model.
```

File: cpp/models/lct_secir/infection_state.h
Comment (annawendler) [inline cpp/models/lct_secir/infection_state.h:245]: Is m_count the number of subcompartments or the number of compartments? The way you use it above it is the number of subcompartments but with this description it is unclear to me what is meant.

File: cpp/models/lct_secir/model.h
Comment (annawendler) [inline cpp/models/lct_secir/model.h:57]:
```suggestion
     * The LCT-SECIR model is defined through ordinary differential equations of the form dydt = f(y, t). 
```

File: cpp/models/lct_secir/model.h
Comment (annawendler) [inline cpp/models/lct_secir/model.h:68]:
```suggestion
     * @brief Calculates the population divided in infection states defined in InfectionStateBase out of a result with subcompartments.
```
What do you mean by "out of a result with subcompartments"?

**Review by annawendler — COMMENTED** [review]
Looks good! I made some comments, mainly regarding notation.

File: cpp/models/lct_secir/infection_state.h
Comment (annawendler) [reply cpp/models/lct_secir/infection_state.h:74]: This is inconsistent in some cases over all files, could you check again?

File: cpp/models/lct_secir/infection_state.h
Comment (lenaploetzke) [reply cpp/models/lct_secir/infection_state.h:54]: You are right, actually never, this only becomes relevant in issue #897 . I will delete this part for this PR.

File: cpp/models/lct_secir/infection_state.h
Comment (lenaploetzke) [reply cpp/models/lct_secir/infection_state.h:245]: I think its better now.

File: cpp/models/lct_secir/model.h
Comment (lenaploetzke) [reply cpp/models/lct_secir/model.h:68]: Rewrote the brief, I think it should be clearer now.

File: cpp/examples/lct_secir.cpp
Comment (annawendler) [reply cpp/examples/lct_secir.cpp:40]: What about this?

Comment (lenaploetzke) [reply cpp/examples/lct_secir.cpp:40]: Forgot to add the file to the commit... thanks! :)

**Review by annawendler — COMMENTED** [review]
No further comments from my side :)

File: cpp/examples/lct_secir.cpp
Comment (lenaploetzke) [reply cpp/examples/lct_secir.cpp]: Yes exactly. You could also call a function to get the number of subcompartments for the infected critical, but this is just an example. :)

File: cpp/examples/lct_secir.cpp
Comment (lenaploetzke) [reply cpp/examples/lct_secir.cpp]: This is also something @HenrZu commented on.

File: cpp/examples/lct_secir.cpp
Comment (lenaploetzke) [reply cpp/examples/lct_secir.cpp:77]: It's actually 30%

**Review by mknaranja — APPROVED** [review]

**Pull request merged**
